library/alloc/src/rc.rs

```diff
@@ -374,19 +374,40 @@ impl<T> Rc<T> {
         }
     }
 
-    /// Constructs a new `Rc<T>` using a closure `data_fn` that has access to a
-    /// weak reference to the constructing `Rc<T>`.
+    /// Constructs a new `Rc<T>` while giving you a `Weak<T>` to the allocation,
+    /// to allow you to construct a `T` which holds a weak pointer to itself.
     ///
     /// Generally, a structure circularly referencing itself, either directly or
-    /// indirectly, should not hold a strong reference to prevent a memory leak.
-    /// In `data_fn`, initialization of `T` can make use of the weak reference
-    /// by cloning and storing it inside `T` for use at a later time.
+    /// indirectly, should not hold a strong reference to itself to prevent a memory leak.
+    /// Using this function, you get access to the weak pointer during the
+    /// initialization of `T`, before the `Rc<T>` is created, such that you can
+    /// clone and store it inside the `T`.
+    ///
+    /// `new_cyclic` first allocates the managed allocation for the `Rc<T>`,
+    /// then calls your closure, giving it a `Weak<T>` to this allocation,
+    /// and only afterwards completes the construction of the `Rc<T>` by placing
+    /// the `T` from your closure inside the allocation.
     ///
     /// Since the new `Rc<T>` is not fully-constructed until `Rc<T>::new_cyclic`
-    /// returns, calling [`upgrade`] on the weak reference inside `data_fn` will
+    /// returns, calling [`upgrade`] on the weak reference inside your closure will
     /// fail and result in a `None` value.
     ///
+    /// # What is the closure argument about?
+    ///
+    /// The point of `new_cyclic` is that the place is allocated first, with
+    /// `strong=0` and `weak=1`. The place has no data yet (you're constructing
+    /// the value to place there), but is in the "zombie" state where it has
+    /// weak references but no data.
+    ///
+    /// # What should the closure return?
+    ///
+    /// The closure returns the object to place in the reference counted place.
+    /// After you've constructed the object, the reference count is "activated"
+    /// and the object is accessable by (clones of) the weak pointer given to the
+    /// constructor closure.
```

Comment:
I think this question and answer could be covered by just adding the word "returned" to "the `T` from your closure" in the third paragraph above:

```diff
- by placing the `T` from your closure inside the allocation.
+ by placing the `T` returned from your closure into the allocation.
```

In general I think we should avoid such 'f.a.q. style' questions in API documentation. If there's obvious questions left after the explanation, then we should find a way to include the missing information there, such that readers are not left with questions that need to be addressed separately.

Reply:
Done as well!